File: src-next/pages/api/patients.js
const database = require('../../lib/database');

export default async function handler(req, res) {
  if (req.method === 'GET') {
    try {
      let { page = 1, limit = 10, search = '' } = req.query;

      let pageNum = parseInt(page);
      let limitNum = parseInt(limit);

      if (isNaN(pageNum) || pageNum < 1) pageNum = 1;
      if (isNaN(limitNum) || limitNum > 100) limitNum = 10;

      const offset = (pageNum - 1) * limitNum;

      let baseQuery = `
        FROM patients p
        LEFT JOIN (
          SELECT mr1.*
          FROM medical_records mr1
          INNER JOIN (
            SELECT patient_id, MAX(visit_date) as max_date
            FROM medical_records
            GROUP BY patient_id
          ) mr2
          ON mr1.patient_id = mr2.patient_id 
          AND mr1.visit_date = mr2.max_date
        ) mr ON p.patient_id = mr.patient_id
        WHERE 1=1
      `;

      const params = [];

      if (search) {
        baseQuery += `
          AND (
            LOWER(p.first_name) LIKE LOWER(?) OR
            LOWER(p.last_name) LIKE LOWER(?) OR
            LOWER(p.patient_id) LIKE LOWER(?)
          )
        `;
        const searchTerm = `%${search}%`;
        params.push(searchTerm, searchTerm, searchTerm);
      }

      const countResult = await database.get(
        `SELECT COUNT(*) as total ${baseQuery}`,
        params
      );

      const patients = await database.query(
        `
        SELECT 
          p.*,
          mr.record_id,
          mr.visit_date,
          mr.chief_complaint
        ${baseQuery}
        ORDER BY p.created_at DESC
        LIMIT ? OFFSET ?
        `,
        [...params, limitNum, offset]
      );

      const today = new Date();

      const processedPatients = patients.map((patient) => {
        let age = null;
        let status = 'New';

        if (patient.date_of_birth) {
          const birthDate = new Date(patient.date_of_birth);
          let calculatedAge = today.getFullYear() - birthDate.getFullYear();
          const m = today.getMonth() - birthDate.getMonth();

          if (m < 0 || (m === 0 && today.getDate() < birthDate.getDate())) {
            calculatedAge--;
          }

          age = calculatedAge;
        }

        if (patient.visit_date) {
          const daysSinceVisit = Math.floor(
            (today - new Date(patient.visit_date)) / (24 * 60 * 60 * 1000)
          );

          if (daysSinceVisit <= 7) status = 'Active';
          else if (daysSinceVisit <= 30) status = 'Scheduled';
          else status = 'Inactive';
        }

        return {
          ...patient,
          id: patient.patient_id,
          age,
          status,
          recent_record: patient.record_id
            ? {
                record_id: patient.record_id,
                visit_date: patient.visit_date,
                chief_complaint: patient.chief_complaint
              }
            : null
        };
      });

      const stats = await database.get(`
        SELECT 
          COUNT(*) as total_patients,
          COUNT(CASE WHEN gender = 'Male' THEN 1 END) as male_patients,
          COUNT(CASE WHEN gender = 'Female' THEN 1 END) as female_patients,
          COUNT(CASE WHEN date_of_birth >= date('now', '-18 years') THEN 1 END) as age_0_18,
          COUNT(CASE WHEN date_of_birth < date('now', '-18 years') AND date_of_birth >= date('now', '-35 years') THEN 1 END) as age_19_35,
          COUNT(CASE WHEN date_of_birth < date('now', '-35 years') AND date_of_birth >= date('now', '-60 years') THEN 1 END) as age_36_60,
          COUNT(CASE WHEN date_of_birth < date('now', '-60 years') THEN 1 END) as age_60_plus,
          COUNT(CASE WHEN created_at >= date('now', '-30 days') THEN 1 END) as new_this_month
        FROM patients
      `);

      res.status(200).json({
        data: processedPatients,
        stats: {
          total_patients: stats.total_patients,
          total_records: stats.total_patients,
          pending_appointments: 0,
          new_this_month: stats.new_this_month,
          age_0_18: stats.age_0_18,
          age_19_35: stats.age_19_35,
          age_36_60: stats.age_36_60,
          age_60_plus: stats.age_60_plus
        },
        pagination: {
          page: pageNum,
          limit: limitNum,
          total: countResult.total,
          totalPages: Math.ceil(countResult.total / limitNum)
        },
        filters: { search }
      });

    } catch (error) {
      console.error({
        message: 'Error fetching patients',
        error: error.message,
        stack: error.stack
      });

      res.status(500).json({
        error: 'Failed to fetch patients'
      });
    }
  }

  // 👉 POST (Create Patient)
  else if (req.method === 'POST') {
    try {
      const data = req.body;

      if (!data.first_name || !data.last_name || !data.date_of_birth || !data.gender) {
        return res.status(400).json({ error: 'Missing required fields' });
      }

      // Generate unique patient_id
      const patientId = 'PAT-' + Math.random().toString(36).substr(2, 9).toUpperCase();

      const result = await database.run(
        `
        INSERT INTO patients (
          patient_id, first_name, last_name, date_of_birth, gender,
          phone, email, address, city, state, zip_code, country,
          blood_type, allergies, medical_history,
          emergency_contact_name, emergency_contact_phone,
          insurance_provider, insurance_policy_number
        )
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        `,
        [
          patientId,
          data.first_name,
          data.last_name,
          data.date_of_birth,
          data.gender,
          data.phone || null,
          data.email || null,
          data.address || null,
          data.city || null,
          data.state || null,
          data.zip_code || null,
          data.country || null,
          data.blood_type || null,
          data.allergies || null,
          data.medical_history || null,
          data.emergency_contact_name || null,
          data.emergency_contact_phone || null,
          data.insurance_provider || null,
          data.insurance_policy_number || null
        ]
      );

      res.status(201).json({ success: true, patient_id: patientId });

    } catch (error) {
      console.error('Error creating patient:', error);
      res.status(500).json({ error: 'Failed to create patient: ' + error.message });
    }
  }

  else {
    res.status(405).json({ error: 'Method not allowed' });
  }
}

File: src-tauri/src/commands/patients.rs
use serde::{Deserialize, Serialize};
use tauri::command;
use sqlx::SqlitePool;
use chrono::{DateTime, Utc};
use anyhow::Result;

use crate::AppState;

#[derive(Debug, Serialize, Deserialize)]
pub struct Patient {
    pub id: String,
    pub first_name: String,
    pub last_name: String,
    pub date_of_birth: String,
    pub gender: Option<String>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub address: Option<String>,
    pub city: Option<String>,
    pub state: Option<String>,
    pub zip_code: Option<String>,
    pub country: Option<String>,
    pub blood_type: Option<String>,
    pub allergies: Option<String>,
    pub medical_history: Option<String>,
    pub emergency_contact_name: Option<String>,
    pub emergency_contact_phone: Option<String>,
    pub insurance_provider: Option<String>,
    pub insurance_policy_number: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[command]
pub async fn get_patients(
    limit: Option<i32>,
    page: Option<i32>,
    search: Option<String>,
    state: tauri::State<'_, AppState>,
) -> Result<Vec<Patient>, String> {
    let pool = state.db_pool.lock().await;
    let pool = pool.as_ref().ok_or("Database not initialized")?;

    let limit = limit.unwrap_or(50);
    let offset = page.unwrap_or(1).saturating_sub(1) * limit;

    let mut query = "SELECT * FROM patients WHERE 1=1".to_string();
    let mut params = Vec::new();

    if let Some(search_term) = search {
        if !search_term.is_empty() {
            query.push_str(" AND (first_name LIKE ? OR last_name LIKE ? OR email LIKE ? OR phone LIKE ?)");
            let search_pattern = format!("%{}%", search_term);
            params.push(search_pattern.clone());
            params.push(search_pattern.clone());
            params.push(search_pattern.clone());
            params.push(search_pattern);
        }
    }

    query.push_str(" ORDER BY last_name, first_name LIMIT ? OFFSET ?");
    params.push(limit.to_string());
    params.push(offset.to_string());

    let mut sql_query = sqlx::query_as::<_, Patient>(&query);
    for param in params {
        sql_query = sql_query.bind(param);
    }

    let results = sql_query
        .fetch_all(pool)
        .await
        .map_err(|e| format!("Failed to fetch patients: {}", e))?;

    Ok(results)
}

#[command]
pub async fn get_patient(
    id: String,
    state: tauri::State<'_, AppState>,
) -> Result<Patient, String> {
    let pool = state.db_pool.lock().await;
    let pool = pool.as_ref().ok_or("Database not initialized")?;

    let result = sqlx::query_as::<_, Patient>(
        "SELECT * FROM patients WHERE id = ?"
    )
    .bind(&id)
    .fetch_one(pool)
    .await
    .map_err(|e| format!("Failed to fetch patient: {}", e))?;

    Ok(result)
}

#[command]
pub async fn create_patient(
    patient: Patient,
    state: tauri::State<'_, AppState>,
) -> Result<Patient, String> {
    let pool = state.db_pool.lock().await;
    let pool = pool.as_ref().ok_or("Database not initialized")?;

    let now = Utc::now().to_rfc3339();

    let new_patient = Patient {
        id: patient.id.clone(),
        first_name: patient.first_name,
        last_name: patient.last_name,
        date_of_birth: patient.date_of_birth,
        gender: patient.gender,
        phone: patient.phone,
        email: patient.email,
        address: patient.address,
        city: patient.city,
        state: patient.state,
        zip_code: patient.zip_code,
        country: patient.country,
        blood_type: patient.blood_type,
        allergies: patient.allergies,
        medical_history: patient.medical_history,
        emergency_contact_name: patient.emergency_contact_name,
        emergency_contact_phone: patient.emergency_contact_phone,
        insurance_provider: patient.insurance_provider,
        insurance_policy_number: patient.insurance_policy_number,
        created_at: now.clone(),
        updated_at: now,
    };

    sqlx::query(
        r#"
        INSERT INTO patients (
            id, first_name, last_name, date_of_birth, gender, phone, email,
            address, city, state, zip_code, country, blood_type, allergies,
            medical_history, emergency_contact_name, emergency_contact_phone,
            insurance_provider, insurance_policy_number, created_at, updated_at
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        "#
    )
    .bind(&new_patient.id)
    .bind(&new_patient.first_name)
    .bind(&new_patient.last_name)
    .bind(&new_patient.date_of_birth)
    .bind(&new_patient.gender)
    .bind(&new_patient.phone)
    .bind(&new_patient.email)
    .bind(&new_patient.address)
    .bind(&new_patient.city)
    .bind(&new_patient.state)
    .bind(&new_patient.zip_code)
    .bind(&new_patient.country)
    .bind(&new_patient.blood_type)
    .bind(&new_patient.allergies)
    .bind(&new_patient.medical_history)
    .bind(&new_patient.emergency_contact_name)
    .bind(&new_patient.emergency_contact_phone)
    .bind(&new_patient.insurance_provider)
    .bind(&new_patient.insurance_policy_number)
    .bind(&new_patient.created_at)
    .bind(&new_patient.updated_at)
    .execute(pool)
    .await
    .map_err(|e| format!("Failed to create patient: {}", e))?;

    Ok(new_patient)
}

#[command]
pub async fn update_patient(
    id: String,
    patient: Patient,
    state: tauri::State<'_, AppState>,
) -> Result<Patient, String> {
    let pool = state.db_pool.lock().await;
    let pool = pool.as_ref().ok_or("Database not initialized")?;

    let now = Utc::now().to_rfc3339();

    let updated_patient = Patient {
        id: id.clone(),
        first_name: patient.first_name,
        last_name: patient.last_name,
        date_of_birth: patient.date_of_birth,
        gender: patient.gender,
        phone: patient.phone,
        email: patient.email,
        address: patient.address,
        city: patient.city,
        state: patient.state,
        zip_code: patient.zip_code,
        country: patient.country,
        blood_type: patient.blood_type,
        allergies: patient.allergies,
        medical_history: patient.medical_history,
        emergency_contact_name: patient.emergency_contact_name,
        emergency_contact_phone: patient.emergency_contact_phone,
        insurance_provider: patient.insurance_provider,
        insurance_policy_number: patient.insurance_policy_number,
        created_at: patient.created_at,
        updated_at: now,
    };

    sqlx::query(
        r#"
        UPDATE patients SET
            first_name = ?, last_name = ?, date_of_birth = ?, gender = ?, phone = ?, email = ?,
            address = ?, city = ?, state = ?, zip_code = ?, country = ?, blood_type = ?, allergies = ?,
            medical_history = ?, emergency_contact_name = ?, emergency_contact_phone = ?,
            insurance_provider = ?, insurance_policy_number = ?, updated_at = ?
        WHERE id = ?
        "#
    )
    .bind(&updated_patient.first_name)
    .bind(&updated_patient.last_name)
    .bind(&updated_patient.date_of_birth)
    .bind(&updated_patient.gender)
    .bind(&updated_patient.phone)
    .bind(&updated_patient.email)
    .bind(&updated_patient.address)
    .bind(&updated_patient.city)
    .bind(&updated_patient.state)
    .bind(&updated_patient.zip_code)
    .bind(&updated_patient.country)
    .bind(&updated_patient.blood_type)
    .bind(&updated_patient.allergies)
    .bind(&updated_patient.medical_history)
    .bind(&updated_patient.emergency_contact_name)
    .bind(&updated_patient.emergency_contact_phone)
    .bind(&updated_patient.insurance_provider)
    .bind(&updated_patient.insurance_policy_number)
    .bind(&updated_patient.updated_at)
    .bind(&id)
    .execute(pool)
    .await
    .map_err(|e| format!("Failed to update patient: {}", e))?;

    Ok(updated_patient)
}

#[command]
pub async fn delete_patient(
    id: String,
    state: tauri::State<'_, AppState>,
) -> Result<(), String> {
    let pool = state.db_pool.lock().await;
    let pool = pool.as_ref().ok_or("Database not initialized")?;

    sqlx::query("DELETE FROM patients WHERE id = ?")
        .bind(&id)
        .execute(pool)
        .await
        .map_err(|e| format!("Failed to delete patient: {}", e))?;

    Ok(())
}

#[command]
pub async fn search_patients(
    query: String,
    limit: Option<i32>,
    state: tauri::State<'_, AppState>,
) -> Result<Vec<Patient>, String> {
    let pool = state.db_pool.lock().await;
    let pool = pool.as_ref().ok_or("Database not initialized")?;

    let limit = limit.unwrap_or(20);
    let search_pattern = format!("%{}%", query);

    let results = sqlx::query_as::<_, Patient>(
        r#"
        SELECT * FROM patients 
        WHERE first_name LIKE ? OR last_name LIKE ? OR email LIKE ? OR phone LIKE ?
        ORDER BY last_name, first_name
        LIMIT ?
        "#
    )
    .bind(&search_pattern)
    .bind(&search_pattern)
    .bind(&search_pattern)
    .bind(&search_pattern)
    .bind(limit)
    .fetch_all(pool)
    .await
    .map_err(|e| format!("Failed to search patients: {}", e))?;

    Ok(results)
}
